### PR TITLE
test: add `driver.clickElementAndWaitForWindowToClose` helper method

### PR DESCRIPTION
**Pull request by davidmurdoch**
There are lots of tests that click a button in our popup window (AKA "MetaMask Dialog") that _Eventually_ close the popup, but they don't _wait_ for the popup to close before continuing on with the rest of the test.

This can cause some race conditions due to the way our test infrastructure finds windows (by title). The problem is that we look for windows of the name "MetaMask Dialog", and in some cases we find 2, we start referencing the first one, then it closes (as it should), and when we try to reference it again Selenium throws (because the window no longer exists).

So this PR introduces a new method "driver.clickElementAndWaitForWindowToClose" that will click the element (just as before), but then wait until the window closes before returning.

## **Related issues**

Fixes:  
https://github.com/MetaMask/metamask-extension/issues/26369
https://github.com/MetaMask/metamask-extension/issues/26353
https://github.com/MetaMask/metamask-extension/issues/26387

### CODE REVIEW

File: test/e2e/webdriver/driver.js [hunk 1]
```diff
@@ -862,6 +862,42 @@ class Driver {
...
+    while (true) {
+      const handles = await this.getAllWindowHandles();
+      if (!handles.includes(handle)) {
+        break;
+      }
+      if (start && Date.now() - start > 15000) {
+        console.info(
+          'The window did not close after 15 seconds, are you sure you closed it?',
+        );
+        start = null; // only log once
```
Comment (HowardBraham) [inline test/e2e/webdriver/driver.js:896]: If the window never closes, this will run until the test times out (usually 80 seconds total).  It also fails with the error
`Error: Timeout of 80000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.`

I don't think that's what we want, I think we want to throw a useful error here.

Comment (davidmurdoch) [reply test/e2e/webdriver/driver.js:896]: There is no error in this method to throw, the test timing out on its own already covers the time out case. This section here is just to help a develop that called this method without understanding that they are responsible for closing the tab.

File: test/e2e/snaps/test-snap-jsx.spec.js [hunk 2]
```diff
@@ -35,8 +34,8 @@ describe('Test Snap JSX', function () {
         await driver.waitForSelector('#connectjsx');
         await driver.clickElement('#connectjsx');
 
-        // switch to metamask extension and click connect
-        await switchToNotificationWindow(driver, 2);
```
Comment (davidmurdoch) [inline test/e2e/snaps/test-snap-jsx.spec.js:39]: `switchToNotificationWindow` is deprecated, so they've been removed as a bit of a drive-by refactor.

File: test/e2e/tests/api-usage/account-tracker-api-usage.spec.ts [hunk 3]
```diff
@@ -1,14 +1,13 @@
 import { strict as assert } from 'assert';
-import { MockedEndpoint } from 'mockttp';
```
Comment (davidmurdoch) [inline test/e2e/tests/api-usage/account-tracker-api-usage.spec.ts:2]: just fixing types in this file.

File: test/e2e/tests/bridge/bridge-button-opens-portfolio.spec.ts [hunk 4]
```diff
@@ -1,21 +1,12 @@
 import { Suite } from 'mocha';
-import { withFixtures, logInWithBalanceValidation } from '../../helpers';
```
Comment (davidmurdoch) [inline test/e2e/tests/bridge/bridge-button-opens-portfolio.spec.ts:2]: just fixing types in this file.

File: test/e2e/tests/bridge/bridge-button-routing.spec.ts [hunk 5]
```diff
@@ -1,20 +1,12 @@
 import { Suite } from 'mocha';
-import { withFixtures, logInWithBalanceValidation } from '../../helpers';
```
Comment (davidmurdoch) [inline test/e2e/tests/bridge/bridge-button-routing.spec.ts:2]: just fixing types in this file.

File: test/e2e/helpers.js [hunk 6]
```diff
@@ -39,6 +39,22 @@ const convertToHexValue = (val) => `0x${new BigNumber(val, 10).toString(16)}`;
 
 const convertETHToHexGwei = (eth) => convertToHexValue(eth * 10 ** 18);
 
+/**
```
Comment (davidmurdoch) [inline test/e2e/helpers.js:42]: it was helpful to have proper types when refactoring to use the new `withWindowUntilClose` method, so I added them here (and fixed several usages of `withFixtures`).

File: test/e2e/webdriver/driver.js (hunk 1 above)
Comment (davidmurdoch) [reply test/e2e/webdriver/driver.js:896]: Maybe we could allow passing a(n optional?) `timeout` value here?

**Review by danjm — DISMISSED** [review]

File: test/e2e/webdriver/driver.js [hunk 7]
```diff
@@ -862,6 +862,47 @@ class Driver {
...
+      }
+
+      const timeElapsed = Date.now() - start;
+      if (timeElapsed > timeout) {
+        throw new Error(
+          `waitForWindowToClose timed out waiting for window handle '${handle}' to close.`,
+        );
+      }
+    }
+  }
```
Comment (HowardBraham) [inline test/e2e/webdriver/driver.js:904]: How do you feel about adding a title?  It would allow for slightly easier-to-read error messages.

```suggestion
  async clickElementAndWaitForWindowToClose(rawLocator, retries = 3) {
    const handle = await this.driver.getWindowHandle();
    const title = await this.driver.getTitle();
    await this.clickElement(rawLocator, retries);
    await this.waitForWindowToClose(handle, title);
  }

  /**
   * Waits for the specified window handle to close before returning.
   *
   * @param {string} handle - The handle of the window or tab we'll wait for.
   * @param {number} [timeout] - The amount of time in milliseconds to wait
   * before timing out. Defaults to `this.timeout`.
   * @throws {Error} throws an error if the window handle doesn't close within
   * the timeout.
   */
  async waitForWindowToClose(handle, title, timeout = this.timeout) {
    const start = Date.now();
    // eslint-disable-next-line no-constant-condition
    while (true) {
      const handles = await this.getAllWindowHandles();
      if (!handles.includes(handle)) {
        return;
      }

      const timeElapsed = Date.now() - start;
      if (timeElapsed > timeout) {
        throw new Error(
          `waitForWindowToClose timed out waiting for window handle '${handle}' with title '${title}' to close.`,
        );
      }
    }
  }
```

Comment (davidmurdoch) [reply test/e2e/webdriver/driver.js:904]: I went down this route to add the title to the error message, but I found it likely to be flaky. The way it's prototyped above, where `title` isn't guaranteed to be the actual title of the window due to race conditions between multiple async calls to query the windows (this async nature of these calls seem to be the cause of another race condition in our tests), seems especially prone to flakiness. 

That said, _all_ of these async window interactions are potentially flaky since the windows can be changing while we are interacting with them. I'm in favor of removing as much potential flakiness as possible from the hot path.

If the async query for the current window title were to be made in the error path of `waitForWindowToClose` it might be slightly better... but a title fetched by `getTitle` isn't _guaranteed_ to match that of the referenced window handle (since the `windowHandle` could finally close during the call to `getTitle`, in which case it would return a different title).

All this just makes me want to switch to a better browser testing framework, if there is one, and/or extend `socket-background.to-mocha.ts` to automatically sync all tab state (`chrome.tabs.query`) with the server, and implement event-driven interactions (e.g., `windowHandle.on("close", ...)`) and maybe even some sort of interaction `lock` mechanism.

**Review by HowardBraham — APPROVED** [review]

**Review by DDDDDanica — APPROVED** [review]

**Review by Mrtenz — APPROVED** [review]
